File: src/Yadasm/Archs/Arch6502.hs
module Yadasm.Archs.Arch6502 where

import           Data.Bits
import qualified Yadasm.Node as N
import qualified Yadasm.Binary as B
import qualified Data.ByteString as ByteString
import qualified Yadasm.Symbol as S
import qualified Yadasm.Line as L
import qualified Yadasm.Context as C
import qualified Yadasm.Comparator as Cmp
import           Text.Printf
import           Data.Maybe (isNothing)
import           Yadasm.Archs.Arch6502Helpers

-- turns an opcode into the desired addressing mode 
-- depending on the cc bits
mask :: InstructionMode -> Integer -> Integer
mask Immediate opcode
  | isCC00 opcode = makeOpcode 0xE0 opcode
  | isCC01 opcode = makeOpcode 0x69 opcode
  | isCC10 opcode = makeOpcode 0xA2 opcode
mask ZeroPage opcode
  | isCC00 opcode = makeOpcode 0xE4 opcode
  | isCC01 opcode = makeOpcode 0x65 opcode
  | isCC10 opcode = makeOpcode 0xA6 opcode
mask ZeroPageX opcode
  | isCC00 opcode = makeOpcode 0xB4 opcode
  | isCC01 opcode = makeOpcode 0x75 opcode
  | isCC10 opcode = makeOpcode 0x75 opcode
mask ZeroPageY opcode
  | isCC10 opcode = makeOpcode 0xB6 opcode
mask Absolute opcode
  | isCC00 opcode = makeOpcode 0xEC opcode
  | isCC01 opcode = makeOpcode 0x6D opcode
  | isCC10 opcode = makeOpcode 0xAE opcode
mask AbsoluteJump opcode
  | isCC00 opcode = makeOpcode 0xEC opcode
mask AbsoluteX opcode
  | isCC00 opcode = makeOpcode 0xBC opcode
  | isCC01 opcode = makeOpcode 0x7D opcode
  | isCC10 opcode = makeOpcode 0x7D opcode
mask AbsoluteY opcode
  | isCC01 opcode = makeOpcode 0x79 opcode
  | isCC10 opcode = makeOpcode 0xBE opcode
mask IndirectX opcode
  | isCC01 opcode = makeOpcode 0x61 opcode
mask IndirectY opcode
  | isCC01 opcode = makeOpcode 0x71 opcode
mask Implied opcode
  | isCC01 opcode = makeOpcode 0x00 opcode
mask Relative opcode
  | isCC00 opcode = opcode
  | isCC01 opcode = opcode
mask Accumulator opcode
  | isCC10 opcode = makeOpcode 0x0A opcode
mask IndirectJump opcode
  | isCC00 opcode = 0x6C -- indirect jmp breaks rules! 
mask Implied opcode = opcode
mask im opcode = maskEcho im opcode

makeImInfo :: String -> InstructionMode -> Integer -> ImInfo
makeImInfo = makeImInfo' mask

makeStore :: String -> Integer -> [ImInfo]
makeStore name opcode =
  [ makeImInfo name ZeroPage opcode
  , makeImInfo name ZeroPageX opcode
  , makeImInfo name Absolute opcode
  , makeImInfo name AbsoluteX opcode
  , makeImInfo name AbsoluteY opcode
  , makeImInfo name IndirectX opcode
  , makeImInfo name IndirectY opcode]

makeLoad :: String -> Integer -> [ImInfo]
makeLoad name opcode = makeImInfo name Immediate opcode:makeStore name opcode

makeAcc :: String -> Integer -> [ImInfo]
makeAcc name opcode = [makeImInfo name Accumulator opcode]

makeLogic :: String -> Integer -> [ImInfo]
makeLogic name opcode = makeAcc name opcode
  ++ [ makeImInfo name ZeroPage opcode
     , makeImInfo name ZeroPageX opcode
     , makeImInfo name Absolute opcode
     , makeImInfo name AbsoluteX opcode]

makeBit :: String -> Integer -> [ImInfo]
makeBit name opcode =
  [makeImInfo name ZeroPage opcode, makeImInfo name Absolute opcode]

makeBranch :: String -> Integer -> [ImInfo]
makeBranch name opcode = [makeImInfo name Relative opcode]

makeImplied :: String -> Integer -> [ImInfo]
makeImplied name opcode = [makeImInfo name Implied opcode]

makeCompareIndex :: String -> Integer -> [ImInfo]
makeCompareIndex name opcode = [ makeImInfo name Immediate opcode
                               , makeImInfo name ZeroPage opcode
                               , makeImInfo name Absolute opcode]

makeDec :: String -> Integer -> [ImInfo]
makeDec name opcode = [ makeImInfo name ZeroPage opcode
                      , makeImInfo name ZeroPageX opcode
                      , makeImInfo name Absolute opcode
                      , makeImInfo name AbsoluteX opcode]

makeJump :: String -> Integer -> [ImInfo]
makeJump name opcode =
  [makeImInfo name AbsoluteJump opcode, makeImInfo name IndirectJump opcode]

makeJsr :: String -> Integer -> [ImInfo]
makeJsr name opcode = [makeImInfo' maskEcho name AbsoluteJump opcode]

makeLoadX :: String -> Integer -> [ImInfo]
makeLoadX name opcode =
  [ makeImInfo name Immediate opcode
  , makeImInfo name ZeroPage opcode
  , makeImInfo name ZeroPageY opcode
  , makeImInfo name Absolute opcode
  , makeImInfo name AbsoluteY opcode]

makeLoadY :: String -> Integer -> [ImInfo]
makeLoadY name opcode =
  [ makeImInfo name Immediate opcode
  , makeImInfo name ZeroPage opcode
  , makeImInfo name ZeroPageX opcode
  , makeImInfo name Absolute opcode
  , makeImInfo name AbsoluteX opcode]

makeStoreX :: String -> Integer -> [ImInfo]
makeStoreX name opcode = [ makeImInfo name ZeroPage opcode
                         , makeImInfo name ZeroPageY opcode
                         , makeImInfo name Absolute opcode]

makeStoreY :: String -> Integer -> [ImInfo]
makeStoreY name opcode = [ makeImInfo name ZeroPage opcode
                         , makeImInfo name ZeroPageX opcode
                         , makeImInfo name Absolute opcode]

nodes' :: ([ImInfo] -> [N.Node]) -> [N.Node]
nodes' makeInstructions = nodes1 ++ nodes2 ++ nodes3
  where
    -- the following node calls are only separated 
    -- because the haskell lsp server did not like such a long function
    nodes1 :: [N.Node]
    nodes1 = makeInstructions
      (makeLoad "adc" 0x69
       ++ makeLogic "asl" 0x1E
       ++ makeLoad "and" 0x21
       ++ makeBit "bit" 0x24
       ++ makeBranch "bpl" 0x10
       ++ makeBranch "bmi" 0x30
       ++ makeBranch "bvc" 0x50
       ++ makeBranch "bvs" 0x70
       ++ makeBranch "bcc" 0x90
       ++ makeBranch "bcs" 0xB0
       ++ makeBranch "bne" 0xD0
       ++ makeBranch "beq" 0xF0
       ++ makeImplied "brk" 0x00
       ++ makeLoad "cmp" 0xC9
       ++ makeCompareIndex "cpx" 0xE0
       ++ makeCompareIndex "cpy" 0xC0
       ++ makeDec "dec" 0xC6
       ++ makeLoad "eor" 0x49
       ++ makeImplied "clc" 0x18)

    nodes2 :: [N.Node]
    nodes2 = makeInstructions
      (makeImplied "sec" 0x38
       ++ makeImplied "cli" 0x58
       ++ makeImplied "sei" 0x78
       ++ makeImplied "clv" 0xB8
       ++ makeImplied "cld" 0xD8
       ++ makeImplied "sed" 0xF8
       ++ makeDec "inc" 0xE6
       ++ makeJump "jmp" 0x4C
       ++ makeJsr "jsr" 0x20
       ++ makeLoad "lda" 0xA9
       ++ makeLoadX "ldx" 0xA2
       ++ makeLoadY "ldy" 0xA0
       ++ makeLogic "lsr" 0x4A
       ++ makeImplied "nop" 0xEA
       ++ makeLoad "ora" 0x09
       ++ makeImplied "tax" 0xAA
       ++ makeImplied "txa" 0x8A
       ++ makeImplied "dex" 0xCA)

    nodes3 :: [N.Node]
    nodes3 = makeInstructions
      (makeImplied "inx" 0xE8
       ++ makeImplied "tay" 0xA8
       ++ makeImplied "dey" 0x88
       ++ makeImplied "iny" 0xC8
       ++ makeLogic "rol" 0x2A
       ++ makeLogic "ror" 0x6A
       ++ makeImplied "rti" 0x40
       ++ makeImplied "rts" 0x60
       ++ makeLoad "sbc" 0xE9
       ++ makeStore "sta" 0x85
       ++ makeImplied "txs" 0x9A
       ++ makeImplied "tsx" 0xBA
       ++ makeImplied "pha" 0x48
       ++ makeImplied "pla" 0x68
       ++ makeImplied "php" 0x08
       ++ makeImplied "plp" 0x28
       ++ makeStoreX "stx" 0x86
       ++ makeStoreY "sty" 0x84)

nodes :: [N.Node]
nodes = nodes' makeInstructions

File: lib/src/archs/arch65/arch6502.rs
use crate::{Node, Word};

use super::{is_cc00, is_cc01, is_cc10, make_opcode, mask_echo, ImInfo, InstModes};

fn mask(im: InstModes, opcode: Word) -> Word {
    match im {
        InstModes::Immediate if is_cc00(opcode) => make_opcode(0xE0, opcode),
        InstModes::Immediate if is_cc01(opcode) => make_opcode(0x69, opcode),
        InstModes::Immediate if is_cc10(opcode) => make_opcode(0xA2, opcode),

        InstModes::ZeroPage if is_cc00(opcode) => make_opcode(0xE4, opcode),
        InstModes::ZeroPage if is_cc01(opcode) => make_opcode(0x65, opcode),
        InstModes::ZeroPage if is_cc10(opcode) => make_opcode(0xA6, opcode),

        InstModes::ZeroPageX if is_cc00(opcode) => make_opcode(0xB4, opcode),
        InstModes::ZeroPageX if is_cc01(opcode) => make_opcode(0x75, opcode),
        InstModes::ZeroPageX if is_cc10(opcode) => make_opcode(0x75, opcode),

        InstModes::ZeroPageY if is_cc10(opcode) => make_opcode(0xEC, opcode),

        InstModes::Absolute if is_cc00(opcode) => make_opcode(0xBC, opcode),
        InstModes::Absolute if is_cc01(opcode) => make_opcode(0x7D, opcode),
        InstModes::Absolute if is_cc10(opcode) => make_opcode(0x7D, opcode),

        InstModes::AbsoluteJump if is_cc00(opcode) => make_opcode(0xEC, opcode),

        InstModes::AbsoluteX if is_cc00(opcode) => make_opcode(0xBC, opcode),
        InstModes::AbsoluteX if is_cc01(opcode) => make_opcode(0x7D, opcode),
        InstModes::AbsoluteX if is_cc10(opcode) => make_opcode(0x7D, opcode),

        InstModes::AbsoluteY if is_cc01(opcode) => make_opcode(0x79, opcode),
        InstModes::AbsoluteY if is_cc10(opcode) => make_opcode(0xBE, opcode),

        InstModes::IndirectX if is_cc01(opcode) => make_opcode(0x61, opcode),
        InstModes::IndirectY if is_cc01(opcode) => make_opcode(0x71, opcode),

        InstModes::Implied if is_cc01(opcode) => make_opcode(0x00, opcode),

        InstModes::Relative if is_cc00(opcode) => mask_echo(im, opcode),
        InstModes::Relative if is_cc01(opcode) => mask_echo(im, opcode),

        InstModes::Accumulator if is_cc10(opcode) => make_opcode(0x0A, opcode),
        // indirect jmp breaks rules
        InstModes::IndirectJump if is_cc00(opcode) => mask_echo(im, 0x6C),

        InstModes::Implied => mask_echo(im, opcode),
        _ => mask_echo(im, opcode),
    }
}

fn make_im_info(im: InstModes, name: &str, opcode: Word) -> ImInfo {
    ImInfo::new(mask, im, name, opcode)
}

fn make_im_from(ims: &[InstModes], name: &str, opcode: Word) -> Vec<ImInfo> {
    ims.iter().map(|i| make_im_info(*i, name, opcode)).collect()
}

fn make_store(name: &str, opcode: Word) -> Vec<ImInfo> {
    make_im_from(
        &[
            InstModes::ZeroPage,
            InstModes::ZeroPageX,
            InstModes::Absolute,
            InstModes::AbsoluteX,
            InstModes::AbsoluteY,
            InstModes::IndirectX,
            InstModes::IndirectY,
        ],
        name,
        opcode,
    )
}

fn make_load(name: &str, opcode: Word) -> Vec<ImInfo> {
    let mut res = make_store(name, opcode);
    res.push(make_im_info(InstModes::Immediate, name, opcode));
    res
}

fn make_acc(name: &str, opcode: Word) -> Vec<ImInfo> {
    vec![make_im_info(InstModes::Accumulator, name, opcode)]
}

fn make_logic(name: &str, opcode: Word) -> Vec<ImInfo> {
    let mut res = make_acc(name, opcode);
    res.append(&mut make_im_from(
        &[
            InstModes::ZeroPage,
            InstModes::ZeroPageX,
            InstModes::Absolute,
            InstModes::AbsoluteX,
        ],
        name,
        opcode,
    ));
    res
}

fn make_bit(name: &str, opcode: Word) -> Vec<ImInfo> {
    make_im_from(&[InstModes::ZeroPage, InstModes::Absolute], name, opcode)
}

fn make_branch(name: &str, opcode: Word) -> Vec<ImInfo> {
    make_im_from(&[InstModes::Relative], name, opcode)
}

fn make_implied(name: &str, opcode: Word) -> Vec<ImInfo> {
    make_im_from(&[InstModes::Implied], name, opcode)
}

fn make_compare_index(name: &str, opcode: Word) -> Vec<ImInfo> {
    make_im_from(
        &[
            InstModes::ZeroPage,
            InstModes::ZeroPageX,
            InstModes::Absolute,
            InstModes::AbsoluteX,
        ],
        name,
        opcode,
    )
}

fn make_dec(name: &str, opcode: Word) -> Vec<ImInfo> {
    make_im_from(
        &[
            InstModes::ZeroPage,
            InstModes::ZeroPageX,
            InstModes::Absolute,
            InstModes::AbsoluteX,
        ],
        name,
        opcode,
    )
}

fn make_jump(name: &str, opcode: Word) -> Vec<ImInfo> {
    make_im_from(
        &[InstModes::AbsoluteJump, InstModes::IndirectJump],
        name,
        opcode,
    )
}

fn make_jsr(name: &str, opcode: Word) -> Vec<ImInfo> {
    make_im_from(&[InstModes::AbsoluteJump], name, opcode)
}

fn make_loadx(name: &str, opcode: Word) -> Vec<ImInfo> {
    make_im_from(
        &[
            InstModes::Immediate,
            InstModes::ZeroPage,
            InstModes::ZeroPageY,
            InstModes::Absolute,
            InstModes::AbsoluteY,
        ],
        name,
        opcode,
    )
}

fn make_loady(name: &str, opcode: Word) -> Vec<ImInfo> {
    make_im_from(
        &[
            InstModes::Immediate,
            InstModes::ZeroPage,
            InstModes::ZeroPageX,
            InstModes::Absolute,
            InstModes::AbsoluteX,
        ],
        name,
        opcode,
    )
}

fn make_storex(name: &str, opcode: Word) -> Vec<ImInfo> {
    make_im_from(
        &[
            InstModes::ZeroPage,
            InstModes::ZeroPageY,
            InstModes::Absolute,
        ],
        name,
        opcode,
    )
}

fn make_storey(name: &str, opcode: Word) -> Vec<ImInfo> {
    make_im_from(
        &[
            InstModes::ZeroPage,
            InstModes::ZeroPageX,
            InstModes::Absolute,
        ],
        name,
        opcode,
    )
}

pub fn make_instructions6502(immediate_size: usize) -> Vec<Node> {
    let mut res = vec![];

    [
        make_load("adc", 0x69),
        make_logic("asl", 0x1E),
        make_load("and", 0x21),
        make_bit("bit", 0x24),
        make_branch("bpl", 0x10),
        make_branch("bmi", 0x30),
        make_branch("bvc", 0x50),
        make_branch("bvs", 0x70),
        make_branch("bne", 0xD0),
        make_branch("beq", 0xF0),
        make_implied("brk", 0x00),
        make_load("cmp", 0xC0),
        make_compare_index("cpx", 0xE0),
        make_compare_index("cpy", 0xC0),
        make_dec("dec", 0xC6),
        make_load("eor", 0x49),
        make_implied("clc", 0x18),
        make_implied("sec", 0x38),
        make_implied("cli", 0x58),
        make_implied("sei", 0x78),
        make_implied("clv", 0xB8),
        make_implied("cld", 0xD8),
        make_implied("sed", 0xF8),
        make_dec("inc", 0xE6),
        make_jump("jmp", 0x4C),
        make_jsr("jsr", 0x20),
        make_load("lda", 0xA9),
        make_loadx("ldx", 0xA2),
        make_loady("ldy", 0xA0),
        make_logic("lsr", 0x4A),
        make_implied("tax", 0xAA),
        make_implied("txa", 0x8A),
        make_implied("dex", 0xCA),
        make_implied("inx", 0xE8),
        make_implied("tay", 0xA8),
        make_implied("dey", 0x88),
        make_implied("iny", 0xC8),
        make_logic("rol", 0x2A),
        make_logic("ror", 0x6A),
        make_implied("rti", 0x40),
        make_implied("rts", 0x60),
        make_load("sbc", 0xE9),
        make_store("sta", 0x85),
        make_implied("txs", 0x9A),
        make_implied("tsx", 0xBA),
        make_implied("pha", 0x48),
        make_implied("pla", 0x68),
        make_implied("php", 0x08),
        make_implied("plp", 0x28),
        make_storex("stx", 0x86),
        make_storey("sty", 0x84),
    ]
    .iter()
    .for_each(|i| {
        res.append(&mut super::arch65helpers::make_instructions(
            i,
            immediate_size,
        ))
    });

    res
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{archs::make_arch, parse_to_strings, Arch, Context};

    fn test_context() -> Context {
        let ctx = Context::new(0x600, 0x700);

        ctx
    }

    fn test_arch() -> Arch {
        make_arch(&make_instructions6502(1), None)
    }

    #[test]
    fn it_should_parse_immediate() {
        let mut ctx = test_context();
        let arch = test_arch();

        let result = parse_to_strings(
            &mut ctx,
            &[
                0xA9, 0xAB, 0xA5, 0xAB, 0xB5, 0xAB, 0xAD, 0x00, 0x02, 0xBD, 0x00, 0x02, 0xB9, 0x00,
                0x02, 0xA1, 0xAb, 0xB1, 0xAB,
            ],
            &[arch],
        );
        let expected = vec![
            "lda #$AB",
            "lda $AB",
            "lda $AB, x",
            "lda $0200",
            "lda $0200, x",
            "lda $0200, y",
            "lda ($AB, x)",
            "lda ($AB), y",
        ]
        .iter()
        .map(|s| s.to_string())
        .collect();

        assert_eq!(Ok(expected), result);
    }
}
